### PR TITLE
Updated the react-collapsing-table to support server driven pagination

### PR DESCRIPTION
**Pull request by JonForest**
I've added three new props to this component:
- currentPage
- totalPages
- updateData (which will also be used to control column sorting in a future PR)

Right now, updateData takes an object `{page: <newPage>}` used by the `nextPage` and `previousPage` functions.
Once I've done something it should also include a sorting key.  Tbd on the format of that.

I've not added any tests for this.

### CODE REVIEW

File: src/components/Table.js [hunk 1]
```diff
@@ -82,15 +82,15 @@ export class Table extends Component {
         this.resizeTable();
     }
 
-    componentWillReceiveProps({ rows }){
+    componentWillReceiveProps({ rows, currentPage, totalPages }){
```
Comment (Salinn) [inline src/components/Table.js]: @JonForest thoughts on refactoring this to something like this?

```
componentWillReceiveProps(props) {
    const {
        rows = [], 
        currentPage = 1, 
        totalPages ((rows.length === 0) ? 1 : Math.ceil(rows.length / currentState.pagination.rowSize))
    } = props
}
```

Comment (JonForest) [reply src/components/Table.js]: Yep, I think we could go with that easily enough.

**Comment by Salinn** [comment]
@JonForest mind if I try making a few changes and I think a logical conclusion is if you pass us the totalPages prop that you are doing server side rendering and we can key off of that.

Some of the changes I was thinking was to add tests and to move the next/last page logic out of this component and ensure that there are some boundary checks for that.

**Comment by JonForest** [comment]
Yeah. Go for it. We've got this implemented already, but we can handle a few minor changes to the API.

**Comment by JonForest** [comment]
We've noticed a bug in this implementation btw.  Will be updating this over the next few days.

**Comment by Salinn** [comment]
@JonForest what was the bug? Any chance you can let me know when you are done fixing the bug? would be happy to help test this branch

**Comment by JonForest** [comment]
@Salinn Bug fixed in the latest commit.  It was not expanding the rows properly on page 2+ when server paginated.

**Comment by alekaizer** [comment]
@Salinn where can it be released, I'm still waiting for it, thought It would have been released this weekend

**Comment by alekaizer** [comment]
@Salinn lemme pull it and see if I can clear up the bug then, I didn't want to mess with Jon's codebase, so I elected to wait for his fix, but I think I should get my hands dirty too

**Comment by alekaizer** [comment]
@Salinn I made the change on #59 but the coverage is still @ -2.3%

**Comment by alekaizer** [comment]
@Salinn I think if we remove the sorting part that the coverage can go to 100%
As @JonForest want to put it in another PR ([here](https://github.com/massmutual/react-collapsing-table/pull/58#issue-234709152))

**Comment by Salinn** [comment]
Thinking about the search functionality and the sorting by columns. Do we need to expand the scope of things here as server side rendering could have more of an effect than just pagination.

File: src/components/Table.js [hunk 2]
```diff
@@ -113,15 +115,28 @@ export class Table extends Component {
     }
 
     nextPage() {
-        this.setState(currentState => {
-            return nextPage({ state: currentState })
-        });
+        if (!this.props.updateData) {
```
Comment (Salinn) [inline src/components/Table.js]: Do we actually need to do this? Isn't the same thing that occurs in the nextPage TableAction?

```
export const nextPage = ({ state }) => {
    const { totalPages, currentPage } = state.pagination;
    const validatedCurrentPage = checkPageState({ newPage: currentPage + 1, totalPages, currentPage });

    return changePage({ state, currentPage: validatedCurrentPage })
};

export const checkPageState = ({ newPage, currentPage, totalPages, shouldCall }) => {
    const isBelowZero = newPage < 0;
    const isZero = newPage === "0";
    const isAboveTotalPages = newPage > totalPages;
    const isNotANumber = isNaN(newPage);
    const isEmpty = newPage.length === 0;

    if(isNotANumber) {
        return currentPage;
    } else if( isZero && shouldCall) {
        return currentPage;
    } else if(isBelowZero) {
        return 1;
    } else if(isAboveTotalPages) {
        return totalPages;
    } else if(isEmpty && shouldCall){
        return currentPage;
    } else {
        return newPage;
    }
};
```

File: src/components/Table.js [hunk 3]
```diff
@@ -48,7 +48,8 @@ export class Table extends Component {
                 rowSize,
                 currentPage,
                 inputtedPage: currentPage,
-                totalPages,
+                totalPages: totalPages || ((rows.length === 0) ? 1 : Math.ceil(rows.length / rowSize)),
+                isServerPagination: totalPages != null
```
Comment (Salinn) [inline src/components/Table.js]: Would it make sense just to set a boolean on the info passed in if  `serverSideRendering` is happening or not? Not just restricting it to pagination?

Comment (alekaizer) [reply src/components/Table.js]: totally agree if the scope is extended to searching and sorting

Comment (Salinn) [reply src/components/Table.js]: I could see the benefit of getting one thing out the door and open up two more prs for each feature set, but maybe change this in support of that as the goal to work towards?

Comment (alekaizer) [reply src/components/Table.js]: Yes, a callback will make it just like CustomPagination

File: src/components/Table.js (hunk 2 above)
Comment (JonForest) [reply src/components/Table.js]: Do you think we should be moving the logic that makes the decision on whether to call the server callback or handle the page change internally to this `TableAction` `nextPage` function, rather than in the `Table.js` itself?
I can see the logic behind that.

File: src/components/Table.js (hunk 3 above)
Comment (JonForest) [reply src/components/Table.js]: My thinking here was the `updateData` prop would be the callback that handles pagination and search via different keys in the argument object.
Right now it just takes `updateData({page: <newPage>})`, but could also take something like 
`updateData({page: <newPage>, ordering: <+/-><column accessor>, searchTerm: <searchText>})`

Examples:
`updateData({page: 2, ordering: '-firstName')` - you want page two of results, ordered reverse alphabetically (Z-a) by firstName
`updateData({ordering: '+firstName', searchTerm: 'bill'})` - you want to search with the term `bill` order order he results a-Z on the firstName column.

The server will have to figure out what to do if you request page 2 on a search term that limits the results to one page, but that seems fine.

Comment (JonForest) [reply src/components/Table.js]: I'm sorry I've dropped off the code coverage on the tests - I wasn't quite sure where to put them and am up against it a little (a lot tbh) on the day-job front.
If we can get the pattern for pagination sorted, I should be able to add ordering and searching fairly easily.

Comment (alekaizer) [reply src/components/Table.js]: @JonForest from my tests, the pagination is sorted out, I have it on prod on 3 differents projects and it's working just fine

**Comment by JonForest** [comment]
@alekaizer have added you commit in here to fix "getting currentpage value when not passed in props"
You can probably remove https://github.com/massmutual/react-collapsing-table/pull/59 so we're just making the changes in one PR
I removed the build file.  That's something we're just using in the Media Suite fork so we can serve from github in the meantime, but shouldn't be in this repo.

**Comment by alekaizer** [comment]
@JonForest oh yeah, will close it

**Comment by Salinn** [comment]
Just checking in. Can I help with writing tests or are you still working on it @alekaizer?

**Comment by JonForest** [comment]
Something very weird has gone on with the linking of our fork to this repo... It's linked itself to another repo, and I'm not entirely sure yet how we fix it.

**Comment by JonForest** [comment]
Okay, I'm going to close this and my colleague will re-raise a new one. Sorry for the delay
